query/src/common/evaluator/scalar.rs

```diff
@@ -0,0 +1,181 @@
+use crate::sql::plans::Scalar;
+use crate::sql::plans::ScalarExpr;
+use crate::sql::IndexType;
+
+enum EvalNode {
+    Function {
+        func: Box<dyn Function>,
+        args: Vec<EvalNode>,
+    },
+    Const {
+        constant: DataValue,
+        data_type: DataTypeImpl,
+    },
+    Variable {
+        index: IndexType,
+    },
+}
+
+pub struct ScalarEvaluator {
+    eval_tree: EvalNode,
+}
+
+impl ScalarEvaluator {
+    pub fn try_create(scalar: &Scalar) -> Result<Self> {
+        let eval_tree = Self::build_eval_tree(scalar)?;
+        Ok(Self { eval_tree })
+    }
+
+    fn build_eval_tree(scalar: &Scalar) -> Result<EvalNode> {
+        match scalar {
+            Scalar::BoundColumnRef(column_ref) => Ok(EvalNode::Variable {
+                index: column_ref.column.index,
+            }),
+            Scalar::ConstantExpr(constant) => Ok(EvalNode::Const {
+                constant: constant.value.clone(),
+                data_type: constant.data_type.clone(),
+            }),
+            Scalar::AndExpr(and) => {
+                let args = vec![
+                    Self::build_eval_tree(&and.left)?,
+                    Self::build_eval_tree(&and.right)?,
+                ];
+                let func = FunctionFactory::instance()
+                    .get("and", &[&and.left.data_type(), &and.right.data_type()])?;
+                Ok(EvalNode::Function { func, args })
+            }
+            Scalar::OrExpr(or) => {
+                let args = vec![
+                    Self::build_eval_tree(&or.left)?,
+                    Self::build_eval_tree(&or.right)?,
+                ];
+                let func = FunctionFactory::instance()
+                    .get("or", &[&or.left.data_type(), &or.right.data_type()])?;
+                Ok(EvalNode::Function { func, args })
+            }
+            Scalar::ComparisonExpr(comp) => {
+                let args = vec![
+                    Self::build_eval_tree(&comp.left)?,
+                    Self::build_eval_tree(&comp.right)?,
+                ];
+                let func = FunctionFactory::instance().get(comp.op.to_func_name(), &[
+                    &comp.left.data_type(),
+                    &comp.right.data_type(),
+                ])?;
+                Ok(EvalNode::Function { func, args })
+            }
+            Scalar::FunctionCall(func) => {
+                let args: Vec<EvalNode> = func
+                    .arguments
+                    .iter()
+                    .map(Self::build_eval_tree)
+                    .collect::<Result<_>>()?;
+                let arg_types: Vec<&DataTypeImpl> = func.arg_types.iter().collect();
+                let func = FunctionFactory::instance().get(func.func_name.as_str(), &arg_types)?;
+                Ok(EvalNode::Function { func, args })
+            }
+            Scalar::Cast(cast) => {
+                let arg = Self::build_eval_tree(&cast.argument)?;
+                let func = CastFunction::create_try(
```

Comment:
Do we have `Scalar::TryCast` ?

1. `cast`  will throw exceptions if an error happens, so its return type is not null.
2. `try_cast`will default to return null if an error happens, its return type is nullable.

Reply:
1. https://docs.snowflake.com/en/sql-reference/functions/try_cast.html
2. https://docs.firebolt.io/sql-reference/functions-reference/try-cast.html

Reply:
We are going to put `Function` object inside `Scalar` later, so every function can be represented with `Scalar::FunctionCall`, including `Cast` and `TryCast`(they can be handled in type checking). 

For now we only keep `TryCast`, since it can satisfy most of the scenarios.